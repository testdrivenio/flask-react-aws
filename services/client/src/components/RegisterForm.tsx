import React from "react";
import { Formik, Form, Field } from "formik";
import { z } from "zod";
import {
  Box,
  Button,
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  VStack,
  Heading,
} from "@chakra-ui/react";
import { Navigate } from "react-router-dom";

interface RegisterFormProps {
  onSubmit: (values: {
    username: string;
    email: string;
    password: string;
  }) => Promise<void>;
  isAuthenticated: () => boolean;
}

const validationSchema = z.object({
  username: z
    .string()
    .min(6, "Username must be at least 6 characters long")
    .regex(
      /^[a-zA-Z0-9_]+$/,
      "Username can only contain letters, numbers, and underscores",
    ),
  email: z
    .string()
    .email("Enter a valid email")
    .min(6, "Email must be at least 6 characters long"),
  password: z.string().min(11, "Password must be at least 11 characters long"),
});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type FormValues = z.infer<typeof validationSchema>;

const RegisterForm: React.FC<RegisterFormProps> = ({
  onSubmit,
  isAuthenticated,
}) => {
  if (isAuthenticated()) {
    return <Navigate to="/" replace />;
  }

  return (
    <Box maxWidth="400px" marginTop={10} margin="100px auto 0">
      <Heading as="h1" size="xl" textAlign="center" mb={6}>
        Register
      </Heading>
      <Formik
        initialValues={{
          username: "",
          email: "",
          password: "",
        }}
        onSubmit={async (values, { setSubmitting, resetForm }) => {
          try {
            await onSubmit(values);
            resetForm();
          } catch (error) {
            console.error("Registration failed:", error);
          } finally {
            setSubmitting(false);
          }
        }}
        validate={(values) => {
          try {
            validationSchema.parse(values);
            return {};
          } catch (error) {
            return (error as z.ZodError).formErrors.fieldErrors;
          }
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form>
            <VStack spacing={4}>
              <Field name="username">
                {({ field }) => (
                  <FormControl
                    isInvalid={!!(errors.username && touched.username)}
                  >
                    <FormLabel htmlFor="username">Username</FormLabel>
                    <Input
                      {...field}
                      id="username"
                      placeholder="Enter a username"
                    />
                    <FormErrorMessage data-testid="errors-username">
                      {errors.username}
                    </FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Field name="email">
                {({ field }) => (
                  <FormControl isInvalid={!!(errors.email && touched.email)}>
                    <FormLabel htmlFor="email">Email</FormLabel>
                    <Input
                      {...field}
                      id="email"
                      type="email"
                      placeholder="Enter an email address"
                    />
                    <FormErrorMessage data-testid="errors-email">
                      {errors.email}
                    </FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Field name="password">
                {({ field }) => (
                  <FormControl
                    isInvalid={!!(errors.password && touched.password)}
                  >
                    <FormLabel htmlFor="password">Password</FormLabel>
                    <Input
                      {...field}
                      id="password"
                      type="password"
                      placeholder="Enter a password"
                    />
                    <FormErrorMessage data-testid="errors-password">
                      {errors.password}
                    </FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Button
                type="submit"
                colorScheme="blue"
                width="full"
                isLoading={isSubmitting}
                bg="green.400"
                _hover={{ bg: "green.500" }}
              >
                Register
              </Button>
            </VStack>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default RegisterForm;
