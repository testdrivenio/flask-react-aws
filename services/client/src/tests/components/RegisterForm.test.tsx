import React from "react";
import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
  act,
} from "../test-utils";
import { describe, it, expect, vi, afterEach } from "vitest";
import { BrowserRouter as Router } from "react-router-dom";
import RegisterForm from "../../components/RegisterForm";
import { expect as expectVitest } from "vitest";

const mockProps = {
  onSubmit: vi.fn(),
  isAuthenticated: vi.fn().mockReturnValue(false),
};

const renderWithRouter = (ui: React.ReactElement, { route = "/" } = {}) => {
  window.history.pushState({}, "Test page", route);
  return render(ui, { wrapper: Router });
};

describe("RegisterForm", () => {
  afterEach(() => {
    cleanup();
  });

  it("renders properly", () => {
    renderWithRouter(<RegisterForm {...mockProps} />);
    const heading = screen.getByRole("heading", { name: "Register" });
    expect(heading.tagName.toLowerCase()).toBe("h1");
  });

  it("renders with default props", () => {
    renderWithRouter(<RegisterForm {...mockProps} />);

    const usernameInput = screen.getByLabelText("Username") as HTMLInputElement;
    expect(usernameInput.value).toBe("");

    const emailInput = screen.getByLabelText("Email") as HTMLInputElement;
    expect(emailInput.type).toBe("email");
    expect(emailInput.value).toBe("");

    const passwordInput = screen.getByLabelText("Password") as HTMLInputElement;
    expect(passwordInput.type).toBe("password");
    expect(passwordInput.value).toBe("");

    const submitButtons = screen.getAllByRole("button", { name: "Register" });
    expect(submitButtons[0].textContent).toBe("Register");
  });

  it("renders", () => {
    const { asFragment } = renderWithRouter(<RegisterForm {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders register form when not authenticated", () => {
    const { container } = renderWithRouter(<RegisterForm {...mockProps} />);
    const heading = container.querySelector("h1");
    expectVitest(heading?.textContent).toBe("Register");
  });

  it("redirects when authenticated", () => {
    const authenticatedProps = {
      ...mockProps,
      isAuthenticated: vi.fn().mockReturnValue(true),
    };
    renderWithRouter(<RegisterForm {...authenticatedProps} />);
    expect(window.location.pathname).toBe("/");
  });
});

describe("handles form validation correctly", () => {
  afterEach(() => {
    cleanup();
  });
  const mockProps = {
    onSubmit: vi.fn(),
    isAuthenticated: vi.fn().mockReturnValue(false),
  };

  it("when fields are empty", async () => {
    renderWithRouter(<RegisterForm {...mockProps} />);

    const submitButton = screen.getByRole("button", { name: "Register" });
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByTestId("errors-username").textContent).to.include(
        "Username must be at least 6 characters long",
      );
      expect(screen.getByTestId("errors-email").textContent).to.include(
        "Enter a valid email",
      );
      expect(screen.getByTestId("errors-password").textContent).to.include(
        "Password must be at least 11 characters long",
      );
    });

    expect(mockProps.onSubmit).not.toHaveBeenCalled();
  });

  it("when email field is not valid", async () => {
    const { getByLabelText, container, findByTestId } = renderWithRouter(
      <RegisterForm {...mockProps} />,
    );

    const form = container.querySelector("form");
    if (!form) throw new Error("Form not found");
    const emailInput = getByLabelText("Email");

    expect(mockProps.onSubmit).toHaveBeenCalledTimes(0);

    await act(async () => {
      fireEvent.change(emailInput, { target: { value: "invalid" } });
      fireEvent.blur(emailInput);
    });

    expect((await findByTestId("errors-email")).innerHTML).toBe(
      "Enter a valid email",
    );

    await act(async () => {
      fireEvent.submit(form);
    });

    await waitFor(() => {
      expect(mockProps.onSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it("when fields are not the proper length", async () => {
    const { getByLabelText, container, findByTestId } = renderWithRouter(
      <RegisterForm {...mockProps} />,
    );

    const form = container.querySelector("form");
    if (!form) throw new Error("Form not found");
    const usernameInput = getByLabelText("Username");
    const emailInput = getByLabelText("Email");
    const passwordInput = getByLabelText("Password");

    expect(mockProps.onSubmit).toHaveBeenCalledTimes(0);

    await act(async () => {
      fireEvent.change(usernameInput, { target: { value: "null" } });
      fireEvent.change(emailInput, { target: { value: "t@t.c" } });
      fireEvent.change(passwordInput, { target: { value: "invalid" } });
      fireEvent.blur(usernameInput);
      fireEvent.blur(emailInput);
      fireEvent.blur(passwordInput);
    });

    expect((await findByTestId("errors-username")).innerHTML).to.include(
      "Username must be at least 6 characters long",
    );
    expect((await findByTestId("errors-email")).innerHTML).to.include(
      "Email must be at least 6 characters long",
    );
    expect((await findByTestId("errors-password")).innerHTML).to.include(
      "Password must be at least 11 characters long",
    );

    await act(async () => {
      fireEvent.submit(form);
    });

    await waitFor(() => {
      expect(mockProps.onSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it("when fields are valid", async () => {
    const { getByLabelText, container } = renderWithRouter(
      <RegisterForm {...mockProps} />,
    );

    const form = container.querySelector("form");
    if (!form) throw new Error("Form not found");
    const usernameInput = getByLabelText("Username");
    const emailInput = getByLabelText("Email");
    const passwordInput = getByLabelText("Password");

    expect(mockProps.onSubmit).toHaveBeenCalledTimes(0);

    await act(async () => {
      fireEvent.change(usernameInput, { target: { value: "proper" } });
      fireEvent.change(emailInput, { target: { value: "t@t.com" } });
      fireEvent.change(passwordInput, { target: { value: "properlength" } });
      fireEvent.blur(usernameInput);
      fireEvent.blur(emailInput);
      fireEvent.blur(passwordInput);

      fireEvent.submit(form);
    });

    await waitFor(() => {
      expect(mockProps.onSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
