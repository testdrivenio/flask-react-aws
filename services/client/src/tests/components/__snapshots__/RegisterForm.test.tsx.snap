// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegisterForm > renders 1`] = `
<DocumentFragment>
  <div
    class="css-vhbdca"
  >
    <h1
      class="chakra-heading css-ov0hf1"
    >
      Register
    </h1>
    <form
      action="#"
    >
      <div
        class="chakra-stack css-1xb2ccj"
      >
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="username"
            id="field-:r6:-label"
          >
            Username
          </label>
          <input
            class="chakra-input css-0"
            id="username"
            name="username"
            placeholder="Enter a username"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="email"
            id="field-:r7:-label"
          >
            Email
          </label>
          <input
            class="chakra-input css-0"
            id="email"
            name="email"
            placeholder="Enter an email address"
            type="email"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <label
            class="chakra-form__label css-uvho5g"
            for="password"
            id="field-:r8:-label"
          >
            Password
          </label>
          <input
            class="chakra-input css-0"
            id="password"
            name="password"
            placeholder="Enter a password"
            type="password"
            value=""
          />
        </div>
        <button
          class="chakra-button css-1lmcr0o"
          type="submit"
        >
          Register
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
